Make native c10d_functional ops work with AOTInductor

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113057
* #113215
* __->__ #113735
* #113438

Summary:
- Revised `c10d_functional` ops to conform to https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native#func
- Modifed `get_cpp_op_schema()` to handle mutable args and aliasing returns


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Commit:
Make native c10d_functional ops work with AOTInductor

[ghstack-poisoned]

chenyang78 commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -66,10 +66,11 @@ c10d::ReduceOp to_reduce_op(const std::string& reduce_op) {
+    at::Tensor& input,
+    std::string reduce_op,
+    // c10::string_view group_name,
+    std::string group_name) {
```
Out of cuorisity, are these function signature changes necessary for making native c10d_functional ops work with AOTInductor? Thanks.

desertfire reviewed (commented):
I am not the best person to review the distributed part.

yifuwang commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -53,6 +59,11 @@ def device(self) -> torch.device:
 
     def _init_process_group(self) -> None:
+        # Allow testing aoti after torch.compile
+        torch._inductor.config.triton.store_cubin = True
```
Sorry I missed this comment earlier. Without this, I'd run into an error if I compile a function with inductor followed by an aot inductor compilation of the same function.

Commit:
Update on "Make native c10d_functional ops work with AOTInductor"


Summary:
- Revised `c10d_functional` ops to conform to https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native#func
- Modifed `get_cpp_op_schema()` to handle mutable args and aliasing returns


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "Make native c10d_functional ops work with AOTInductor"


Summary:
- Revised `c10d_functional` ops to conform to https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native#func
- Modifed `get_cpp_op_schema()` to handle mutable args and aliasing returns


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
